File: javascript/libjoynr-js/src/test/js/joynr/messaging/websocket/WebSocketMessagingSkeletonTest.js
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
require("../../../node-unit-test-helper");
const WebSocketMessagingSkeleton = require("../../../../../main/js/joynr/messaging/websocket/WebSocketMessagingSkeleton");
const JoynrMessage = require("../../../../../main/js/joynr/messaging/JoynrMessage");
const JSONSerializer = require("../../../../../main/js/joynr/util/JSONSerializer");
const MessageSerializer = require("../../../../../main/js/joynr/messaging/MessageSerializer");
const MulticastPublication = require("../../../../../main/js/joynr/dispatching/types/MulticastPublication");
const Request = require("../../../../../main/js/joynr/dispatching/types/Request");
const WebSocketAddress = require("../../../../../main/js/generated/joynr/system/RoutingTypes/WebSocketAddress");
const WebSocketClientAddress = require("../../../../../main/js/generated/joynr/system/RoutingTypes/WebSocketClientAddress");
const SharedWebSocket = require("../../../../../main/js/joynr/messaging/websocket/SharedWebSocket");

describe("libjoynr-js.joynr.messaging.websocket.WebSocketMessagingSkeleton", () => {
    let window = null;
    let sharedWebSocket = null;
    let webSocketMessagingSkeleton = null;
    let listener = null;
    let data = null;
    let event = null;
    let multicastEvent;
    const ccAddress = new WebSocketAddress({
        protocol: "ws",
        host: "host",
        port: 1234,
        path: "/test"
    });
    const localAddress = new WebSocketClientAddress({
        id: "1234"
    });

    beforeEach(done => {
        function Window() {}
        window = new Window();
        window.addEventListener = jasmine.createSpy("addEventListener");

        sharedWebSocket = new SharedWebSocket({
            remoteAddress: ccAddress,
            localAddress
        });

        spyOn(sharedWebSocket, "send").and.callThrough();

        webSocketMessagingSkeleton = new WebSocketMessagingSkeleton({
            sharedWebSocket,
            mainTransport: true
        });

        listener = jasmine.createSpy("listener");
        function MessageEvent() {}
        event = new MessageEvent();
        const request = Request.create({
            methodName: "methodName"
        });
        const joynrMessage = new JoynrMessage({
            type: JoynrMessage.JOYNRMESSAGE_TYPE_REQUEST,
            payload: JSON.stringify(request)
        });
        joynrMessage.from = "fromParticipantId";
        joynrMessage.to = "toParticipantId";
        joynrMessage.expiryDate = Date.now() + 10000;
        data = joynrMessage;
        multicastEvent = new MessageEvent();
        if (typeof Buffer === "function") {
            // node environment
            event.data = MessageSerializer.stringify(joynrMessage);
            event.target = {
                binaryType: "arraybuffer"
            };
            const multicastId = "multicastId";
            const publication = MulticastPublication.create({
                response: ["test"],
                multicastId
            });

            const multicastJoynrMessage = new JoynrMessage({
                type: JoynrMessage.JOYNRMESSAGE_TYPE_MULTICAST,
                payload: JSONSerializer.stringify(publication)
            });
            multicastJoynrMessage.from = "fromParticipantId";
            multicastJoynrMessage.to = "toParticipantId";
            multicastJoynrMessage.expiryDate = Date.now() + 10000;
            multicastEvent.data = MessageSerializer.stringify(multicastJoynrMessage);
        } else if (typeof TextEncoder === "function") {
            // browser
            const textEncoder = new TextEncoder();
            event.data = textEncoder.encode(JSON.stringify(data));
            event.target = {
                binaryType: "arraybuffer"
            };
            multicastEvent.data = textEncoder.encode(
                JSON.stringify(
                    new JoynrMessage({
                        type: JoynrMessage.JOYNRMESSAGE_TYPE_MULTICAST
                    })
                )
            );
        } else {
            // browser without TextDecoder
            done.fail("error in beforeEach: Buffer/TextEncoder not supported");
        }
        done();
    });

    it("is of correct type and has all members", done => {
        expect(WebSocketMessagingSkeleton).toBeDefined();
        expect(typeof WebSocketMessagingSkeleton === "function").toBeTruthy();
        expect(webSocketMessagingSkeleton).toBeDefined();
        expect(webSocketMessagingSkeleton instanceof WebSocketMessagingSkeleton).toBeTruthy();
        expect(webSocketMessagingSkeleton.registerListener).toBeDefined();
        expect(typeof webSocketMessagingSkeleton.registerListener === "function").toBeTruthy();
        expect(webSocketMessagingSkeleton.unregisterListener).toBeDefined();
        expect(typeof webSocketMessagingSkeleton.unregisterListener === "function").toBeTruthy();
        done();
    });

    it("throws if arguments are missing or of wrong type", done => {
        expect(() => {
            webSocketMessagingSkeleton = new WebSocketMessagingSkeleton({
                sharedWebSocket,
                mainTransport: true
            });
        }).not.toThrow(); // correct call
        expect(() => {
            webSocketMessagingSkeleton = new WebSocketMessagingSkeleton({
                sharedWebSocket
            });
        }).toThrow(); // mainTransport missing
        expect(() => {
            webSocketMessagingSkeleton = new WebSocketMessagingSkeleton({});
        }).toThrow(); // incorrect call

        expect(() => {
            webSocketMessagingSkeleton.registerListener(() => {});
        }).not.toThrow(); // correct call
        expect(() => {
            webSocketMessagingSkeleton.registerListener("");
        }).toThrow(); // listener is of wrong type
        expect(() => {
            webSocketMessagingSkeleton.registerListener({});
        }).toThrow(); // listener is of wrong type

        expect(() => {
            webSocketMessagingSkeleton.unregisterListener(() => {});
        }).not.toThrow(); // correct call
        expect(() => {
            webSocketMessagingSkeleton.unregisterListener("");
        }).toThrow(); // listener is of wrong type
        expect(() => {
            webSocketMessagingSkeleton.unregisterListener({});
        }).toThrow(); // listener is of wrong type
        done();
    });

    it("event calls through to registered listener", () => {
        webSocketMessagingSkeleton.registerListener(listener);
        expect(listener).not.toHaveBeenCalled();

        sharedWebSocket.onmessage(event);

        expect(listener.calls.argsFor(0)[0].id).toBe(data.id);
        expect(listener.calls.argsFor(0)[0].from).toBe(data.from);
        expect(listener.calls.argsFor(0)[0].to).toBe(data.to);
        expect(listener.calls.count()).toBe(1);
    });

    it("event does not call through to unregistered listener", () => {
        webSocketMessagingSkeleton.registerListener(listener);
        sharedWebSocket.onmessage(event);
        expect(listener).toHaveBeenCalled();
        expect(listener.calls.count()).toBe(1);
        webSocketMessagingSkeleton.unregisterListener(listener);
        sharedWebSocket.onmessage(event);
        expect(listener.calls.count()).toBe(1);
    });

    function receiveMessageAndCheckForIsReceivedFromGlobalFlag(expectedValue) {
        webSocketMessagingSkeleton.registerListener(listener);

        sharedWebSocket.onmessage(multicastEvent);

        expect(listener).toHaveBeenCalled();
        expect(listener.calls.count()).toBe(1);
        expect(listener.calls.argsFor(0)[0].isReceivedFromGlobal).toBe(expectedValue);
    }

    it("sets isReceivedFromGlobal if web socket is main transport", () => {
        receiveMessageAndCheckForIsReceivedFromGlobalFlag(true);
    });

    it("does not set isReceivedFromGlobal if web socket is NOT main transport", () => {
        webSocketMessagingSkeleton = new WebSocketMessagingSkeleton({
            sharedWebSocket,
            mainTransport: false
        });

        receiveMessageAndCheckForIsReceivedFromGlobalFlag(false);
    });
});

File: java/messaging/websocket/joynr-websocket-common/src/test/java/io/joynr/messaging/websocket/WebSocketMessagingSkeletonTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2021 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.messaging.websocket;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.HashSet;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.runners.MockitoJUnitRunner;

import com.google.inject.name.Names;

import io.joynr.messaging.FailureAction;
import io.joynr.messaging.JoynrMessageProcessor;
import io.joynr.messaging.routing.MessageRouter;
import io.joynr.messaging.websocket.WebSocketMessagingSkeleton.MainTransportFlagBearer;
import joynr.ImmutableMessage;
import joynr.Message.MessageType;
import joynr.MutableMessage;
import joynr.system.RoutingTypes.WebSocketAddress;

@RunWith(MockitoJUnitRunner.class)
public class WebSocketMessagingSkeletonTest {
    private static final FailureAction NO_FAILURE_EXPECTED = new FailureAction() {
        @Override
        public void execute(Throwable error) {
            fail("Unexpected FailureAction call: " + error.toString());
        }
    };

    private static final String RECIPIENT_ID = "recipientId";
    private static final String SENDER_ID = "senderId";

    private static byte[] createMessage(MessageType type) throws Exception {
        MutableMessage mutableMessage = new MutableMessage();
        mutableMessage.setPayload("testPayload".getBytes());
        mutableMessage.setRecipient(RECIPIENT_ID);
        mutableMessage.setSender(SENDER_ID);
        mutableMessage.setType(type);
        return mutableMessage.getImmutableMessage().getSerializedMessage();
    }

    private WebSocketMessagingSkeleton subject;

    @Mock
    private MainTransportFlagBearer mainTransportFlagBearer;

    @Mock
    private MessageRouter messageRouter;

    @Before
    public void setUp() throws Exception {
        // create test subject
        subject = new WebSocketMessagingSkeleton(Mockito.mock(WebSocketAddress.class),
                                                 Mockito.mock(WebSocketEndpointFactory.class),
                                                 messageRouter,
                                                 mainTransportFlagBearer,
                                                 new HashSet<JoynrMessageProcessor>());

    }

    private void verifyReceivedFromGlobal(boolean expected) throws Exception {
        for (MessageType type : MessageType.values()) {
            subject.transmit(createMessage(type), NO_FAILURE_EXPECTED);
            ArgumentCaptor<ImmutableMessage> messageCaptor = ArgumentCaptor.forClass(ImmutableMessage.class);
            verify(messageRouter).routeIn(messageCaptor.capture());
            assertEquals(expected, messageCaptor.getValue().isReceivedFromGlobal());
            reset(messageRouter);
        }
    }

    @Test
    public void doNotSetReceivedFromGlobalInCC() throws Exception {
        verifyReceivedFromGlobal(false);
    }

    @Test
    public void doSetReceivedFromGlobalInApp() throws Exception {
        when(mainTransportFlagBearer.isMainTransport()).thenReturn(true);
        setUp(); //Re-initialize subject
        verifyReceivedFromGlobal(true);
    }

}
